src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -880,7 +880,7 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
   LOG_PREFIX(ObjectDataHandler::prepare_data_reservation);
   ceph_assert(size <= max_object_size);
   if (!object_data.is_null()) {
-    ceph_assert(object_data.get_reserved_data_len() == max_object_size);
+    ceph_assert(object_data.get_reserved_data_len() >= size);
```

Comment:
I'm actually not following the concepts of:
* onode_layout_t::size
* onode_layout_t::object_data::reserved_data_len
* the configuration seastore_default_max_object_size
* ObjectDataHandler::max_object_size
* the expected_object_size from OP_SETALLOCHINT

And how they should interact correctly with each other in this PR.

@athanatos @xxhdx1985126 do you have more inputs ?

Reply:
@athanatos We intended to avoid the ObjectDataHandler's over-allocating issue, because it looks to us that this laddr over-allocating may lead to the exhaustion of logical address space, am I right?. We thought that the `expected_object_size` can be treated as the upper bound of objects' size. Now that this expectation is wrong, is there any other way to let OSD know the appropriate reservation space size? Thanks.

Reply:
IIRC the final solution to avoid the exhaustion of logical address space is to extend it beyond 64 bits.

> is there any other way to let OSD know the appropriate reservation space size?

Does ObjectStore have a limitation of max object size supported? If not, probably we can change the configuration `seastore_default_max_object_size` to `seastore_default_object_data_reservation` (as well as the according logics) for this goal? We will also need to handle the case when the actual object data size passes the reserved size.

Reply:
> IIRC the final solution to avoid the exhaustion of logical address space is to extend it beyond 64 bits.

Great!!! Hah:D We are thinking the same thing:-) We are designing a 128bit lba key format and may be able to present a document this week, from our opinion, with a 128 bit lba key, we can also have a series of other benefits, including significantly reducing the overhead of "OP_CLONE" and better [LBC](https://docs.google.com/document/d/1TOZ1jnlxIKwmp3CE7Vr7vE1vujergXa9) performance.

For example, if we put "snap id" in higher bits, we can put clone objects in different address spaces, avoiding the lba key collision when allocating lba mappings for cloned objects:-)

Reply:
Yeah, 128 bit key is probably the logical step.  There's an osd_max_object_size configurable set to 128M.  In practice, we can probably require a smaller max object size.